felix/dataplane/linux/qos_controls.go

```diff
@@ -18,138 +18,74 @@ import (
 	"errors"
 	"fmt"
 
-	"github.com/vishvananda/netlink"
-
-	"github.com/projectcalico/calico/felix/dataplane/linux/bwp"
+	"github.com/projectcalico/calico/felix/dataplane/linux/qos"
 	"github.com/projectcalico/calico/felix/proto"
 )
 
 func (m *endpointManager) maybeUpdateQoSBandwidth(old, new *proto.WorkloadEndpoint) error {
 	var errs []error
+
+	var oldName, newName string
 	var oldIngressBw, oldEgressBw, newIngressBw, newEgressBw int64
 	var oldIngressBurst, oldEgressBurst, newIngressBurst, newEgressBurst int64
-	if old != nil && old.QosControls != nil {
-		oldIngressBw = old.QosControls.IngressBandwidth
-		oldEgressBw = old.QosControls.EgressBandwidth
-		oldIngressBurst = old.QosControls.IngressBurst
-		oldEgressBurst = old.QosControls.EgressBurst
+
+	if old != nil {
+		oldName = old.Name
+		if old.QosControls != nil {
+			oldIngressBw = old.QosControls.IngressBandwidth
+			oldEgressBw = old.QosControls.EgressBandwidth
+			oldIngressBurst = old.QosControls.IngressBurst
+			oldEgressBurst = old.QosControls.EgressBurst
+		}
 	}
 
-	if new != nil && new.QosControls != nil {
-		newIngressBw = new.QosControls.IngressBandwidth
-		newEgressBw = new.QosControls.EgressBandwidth
-		newIngressBurst = new.QosControls.IngressBurst
-		newEgressBurst = new.QosControls.EgressBurst
+	if new != nil {
+		newName = new.Name
+		if new.QosControls != nil {
+			newIngressBw = new.QosControls.IngressBandwidth
+			newEgressBw = new.QosControls.EgressBandwidth
+			newIngressBurst = new.QosControls.IngressBurst
+			newEgressBurst = new.QosControls.EgressBurst
+		}
 	}
 
-	if oldIngressBw != newIngressBw || oldIngressBurst != newIngressBurst {
-		if oldIngressBw != 0 {
-			err := removeIngressQdisc(old.Name)
+	if oldName != newName || oldIngressBw != newIngressBw || oldIngressBurst != newIngressBurst {
+		if oldName == newName && oldIngressBw != 0 && newIngressBw != 0 {
```

Comment:
This logic seems to assume that the old values match the dataplane; we try not to make such assumptions in Felix.  So what I'd expect to see is logic that reads back the dataplane state, compares it with the desired state and then "fixes" the dataplane by adding/removing/changing the qdisc.

I'd probabaly use a DeltaTracker for this with a struct that tracks the state of the dataplane:
```
type qosState struct {
  IngressBandwidth int
  IngressBurst int
}
```
then you need to be able to read back those values from the dataplane to populate the "Dataplane()` side of the delta tracker, and populate the "Desired()" side with  the values from the datastore.

Then the delta tracker has methods for iterating over all the updates/additions/deletions that are needed.